Init Creating Vincent Tool doc (#130)
LIT-Protocol/Vincent: packages/sdk/docs/Developers/Creating-Tools.md

---
category: Developers
title: Creating Vincent Tools
---

# What is a Vincent Tool?

> A Vincent Tool is a function built using [Lit Actions](https://developer.litprotocol.com/sdk/serverless-signing/overview) that enables Vincent Apps to perform specific actions on behalf of Vincent App Users. These tools are the core functional units that Vincent Apps use to interact with blockchains, APIs, and other services while being governed by user-configured Vincent Policies.

Reviewer: Consider:

```suggestion
A Vincent Tool is a function built using [Lit Actions](https://developer.litprotocol.com/sdk/serverless-signing/overview) that enables Vincent Apps to perform specific actions on behalf of Vincent App Users. Tools are the core functional units that Vincent Apps use to interact with blockchains, APIs, and other services while being governed by user-configured [Vincent Policies](./Creating-Policies.md).
```

Reviewer: Even better! :)

## Key Capabilities

**Flexible Data Access**
- Read and write on-chain data (token balances, NFT ownership, smart contract state)
- Read and write off-chain data from any HTTP-accessible API or database

**Policy-Driven Execution**
- Execute tools only when all Vincent App User registered Vincent Policies allow it
- Combine on and off-chain policy conditions

**Type-Safe Development**
- Strongly-typed Zod schemas for your tool functions' parameters and return values
- Clear interfaces between your tool and Vincent Policies with type safety

**Lit Action Based**
- Built-in ethers.js for blockchain interactions
- Ability to import and use NPM packages within your tool
- Access to Lit Protocol's [Encryption and Access Control](https://developer.litprotocol.com/sdk/access-control/intro) features

> - Ability to sign data on behalf of the Vincent App User's Agent Wallet

Reviewer: Maybe we should provide a link here? It seems the definition for "Agent Wallet" is in PR #105 though

- **Custom Cryptographic Protocols**: Implement specialized signing schemes or multi-party computation workflows

**Programmable Logic & Governance**
- **Conditional Execution**: Implement complex business logic with branching paths based on runtime data
- **Policy Integration**: Leverage Vincent App User configured policies to govern when and how Vincent Tools can execute
- **Stateful Operations**: Maintain and update persistent state across multiple Vincent Tool executions

# How a Vincent Tool Works

A Vincent Tool consists of two main lifecycle methods executed in the following order:

1. **Precheck**: Executed locally by the Vincent Tool executor, this function provides a best-effort check that the tool execution shouldn't fail
   - Before the execution of your tool's `precheck` function, the Vincent Tool & Policy SDK will execute the `precheck` functions of the Vincent Policies enabled by the Vincent App User for your tool for the specific Vincent App the tool is being executed for
   - If all Vincent Policies return `allow` results, the Vincent Tool's `precheck` function will be executed

>    - This function is where you'd perform checks such as validating the Vincent Agent Wallet has enough balance to execute the tool logic, has the appropriate on-chain approvals to make token transfers, or anything else your tool can validate before executing the tool's logic

Author: Changed to:

- This function is where you'd perform checks such as validating the Vincent Agent Wallet has sufficient token balances, has the appropriate on-chain approvals to make token transfers, or anything else your tool can validate before executing the tool's logic

2. **Execute**: Executed within the Lit Action environment, this function performs the actual tool logic and has the ability to sign data using the Vincent App User's Agent Wallet

>    - Before the execution of your tool's `execute` function, the Vincent Tool & Policy SDK will execute the `evaluate` functions of the Vincent Policies enabled by the Vincent App User for your tool for the specific Vincent App the tool is being executed for

Reviewer: Similar to what I noted above, I think it's implied that the policies in question here are specific to the tool
```suggestion
   - Before the execution of your tool's `execute` function, the Vincent Tool & Policy SDK will execute the `evaluate` functions of the Vincent Policies
```

      };
    };
    deniedPolicy?: {
      policyPackageName: string;
      result: evalDenyResultSchema;
    };
  };
  succeed: (executeSuccessResult: executeSuccessSchema) => void;
  fail: (executeFailResult: executeFailSchema) => void;
}
```

Where:
- `toolIpfsCid`: The IPFS CID of the Vincent Tool that is being executed
- `appId`: The ID of the Vincent App the Vincent Tool is being executed for
- `appVersion`: The version of the Vincent App the Vincent Tool is being executed for
- `delegation`:
  - `delegateeAddress`: The Ethereum address of the Vincent Tool executor
  - `delegatorPkpInfo`:
    - `tokenId`: The token ID of the Vincent App User's Vincent Agent Wallet
    - `ethAddress`: The Ethereum address of the Vincent App User's Vincent Agent Wallet
    - `publicKey`: The public key of the Vincent App User's Vincent Agent Wallet
- `policiesContext`: An object containing the context of the Vincent Policies enabled by the Vincent App User for your tool for the specific Vincent App the tool is being executed for
  - `allow`: A boolean indicating if the Vincent Tool execution is allowed to proceed, and all evaluated Vincent Policies returned `allow` results

>   - `allowedPolicies`: An object containing the results of the `evaluate` functions of the Vincent Policies enabled by the Vincent App User for your tool for the specific Vincent App the tool is being executed for

Reviewer: Consider changing this to
```suggestion
  - `allowedPolicies`: An object containing the results of the `evaluate` functions of the Vincent Policies
```

Author: Changed to:

- `allowedPolicies`: An object containing the results and `commit` functions for each Vincent Policy that permitted tool execution

>     - `[policyPackageName]`: An object where the key is the package name of the Vincent Policy, and the value is an object containing the result of the `evaluate` function of the Vincent Policy as well the policy's `commit` function if it exists

Author: Changed to:

- `[policyPackageName]`: An object where the key is the package name of the Vincent Policy, and the value is an object containing the result of the policy's `evaluate` function and the policy's `commit` function if it exists

        - The parameter object passed to the `commit` function is defined by each Vincent Policy's [commitParamsSchema](./Creating-Policies.md#commitparamsschema), and the return value is defined by the policy's [commitAllowResultSchema](./Creating-Policies.md#commitallowresultschema) or [commitDenyResultSchema](./Creating-Policies.md#commitdenyresultschema)
  - `deniedPolicy`: An object containing the first Vincent Policy that denied the Vincent Tool execution
    - `policyPackageName`: The package name of the Vincent Policy that denied the Vincent Tool execution
    - `result`: The result of the `evaluate` function of the Vincent Policy that denied the Vincent Tool execution, will have the shape of the Vincent Policy's [evalDenyResultSchema](./Creating-Policies.md#evaldenyresultschema)
- `succeed`: A helper method for returning a `success` result from your tool's `precheck` and `execute` functions
- `fail`: A helper method for returning a `fail` result from your tool's `precheck` and `execute` functions

## Parameter Schemas

### `toolParamsSchema`

> This Zod schema defines the structure of parameters that executors of your tool will provide to your tool. These should be the parameters you require to execute your tool's functionality, as well as any parameters required by the Vincent Policies your tool supports.

Reviewer: Consider this
```suggestion
This Zod schema defines the structure of parameters that executors will provide to your tool. These should be the parameters you require to execute your tool's functionality, as well as any parameters required by the Vincent Policies your tool supports.
```

For example, if you are building a token transfer tool that supports a Vincent spending limit policy, you might define the `toolParamsSchema` as follows:

```typescript
import { createVincentTool } from '@lit-protocol/vincent-tool-sdk';
import { z } from 'zod';

const toolParamsSchema = z.object({
  tokenAddress: z.string(),
  amountToSend: z.number(),
  recipientAddress: z.string(),
});

const vincentTool = createVincentTool({
  // ... other tool definitions

  toolParamsSchema,
});
```

These parameters give your tool what it needs to send a transaction transferring `amountToSend` of `tokenAddress` to `recipientAddress`.

The `tokenAddress` and `amountToSend` parameters are also the parameters required to be given to the Vincent spending limit policy which we'll cover in the next section.

## Defining Supported Policies

> To add support for Vincent Policies to your tool, you need to create _VincentToolPolicy_ objects using the `createVincentToolPolicy` function from the `@lit-protocol/vincent-tool-sdk` package for each Vincent Policy you want your tool to support. These _VincentToolPolicy_ objects are then added to your tool's `supportedPolicies` array, which binds the policies to your tool and enables proper parameter mapping between your tool and the policies.

Reviewer: Should VincentToolPolicy be italicized?

Author: Ideally it would link the to the typedoc interface for reference. I'll add a TODO to update once the typedocs are published

  toolParameterMappings: {
    tokenAddress: 'tokenAddress',
    amountToSend: 'amount',
  },
});

const vincentTool = createVincentTool({
  // ... other tool definitions

  toolParamsSchema,

  supportedPolicies: supportedPoliciesForTool([SpendingLimitPolicy]),
});
```

A couple of new things are happening in this code example:

First we're importing `bundledVincentPolicy` from the `@lit-protocol/vincent-policy-spending-limit` package, which is a Vincent Policy object created using the Vincent Tool & Policy SDK and exported by the policy author for Vincent Tools to consume

Then we're creating a `VincentToolPolicy` object named `SpendingLimitPolicy` using the `createVincentToolPolicy` function. The `createVincentToolPolicy` function takes a single object parameter with the required properties:

- `toolParamsSchema`: The Zod schema (covered in the [Parameter Schemas](#parameter-schemas) section) you've defined for the parameters your tool expects to be given by the Vincent Tool executor
- `bundledVincentPolicy`: The Vincent Policy object created by the policy author for Vincent Tools to consume, which is imported from the `@lit-protocol/vincent-policy-spending-limit` package

> - `toolParameterMappings`: An object that maps the parameters given to your tool, to the parameters expected by the Vincent Policy you're supporting

Reviewer: Consider removing
```suggestion
- `toolParameterMappings`: An object that maps the parameters given to your tool to the parameters expected by the Vincent Policy you're supporting
```

  - The keys of this object are the parameter names your tool uses (`tokenAddress` and `amountToSend`), and the values are the parameter names expected by the Vincent Policy (`tokenAddress` and `amount`)

Lastly, we take the `SpendingLimitPolicy` object and add it to an array, which we then wrap in a `supportedPoliciesForTool` function call to our tool's `supportedPolicies` array.

> This is how we register the `SpendingLimitPolicy` with our tool, and is all that's needed for your tool to support the Vincent spending limit policy. The execution of the policy's `precheck` and `evaluate` functions will be handled for you by the Vincent Tool & Policy SDK, as well as processing the return values from the policy's `precheck` and `evaluate` functions to check if the tool should be allowed to execute.

Reviewer: Consider
```suggestion
This is how we register the `SpendingLimitPolicy` with our tool. That's all that's needed for your tool to support the Vincent spending limit policy. The execution of the policy's `precheck` and `evaluate` functions will be handled for you by the Vincent Tool & Policy SDK, as well as processing the return values from the policy's `precheck` and `evaluate` functions to check if the tool should be allowed to execute.
```

## Precheck Function

The `precheck` function is executed locally by the Vincent Tool executor to provide a best-effort check that the tool execution shouldn't fail when the `execute` function is called.

Executing a Vincent Tool's `execute` function uses the Lit network, which costs both time and money, so your `precheck` function should perform whatever validation it can to ensure that the tool won't fail during execution.

> Before executing your tool's `precheck` function, the Vincent Tool & Policy SDK will execute the `precheck` functions of the Vincent Policies enabled by the Vincent App User for your tool the specific Vincent App the tool is being executed for. If all Vincent Policies return `allow` results, the Vincent Tool's `precheck` function will be executed.

Reviewer: Consider reducing this to
```suggestion
Before executing your tool's `precheck` function, the Vincent Tool & Policy SDK will execute the `precheck` functions of the Vincent Policies. If all Vincent Policies return `allow` results, the Vincent Tool's `precheck` function will be executed.
```

The following schema returns useful information to the Vincent Tool executor about the current balances of the Agent Wallet, as well as the estimated gas cost of the transaction:

```typescript
import { createVincentTool } from '@lit-protocol/vincent-tool-sdk';
import { z } from 'zod';

const vincentTool = createVincentTool({
  // ... other tool definitions

  precheckSuccessSchema: z.object({
    erc20TokenBalance: z.number(),
    nativeTokenBalance: z.number(),
    estimatedGas: z.number(),
  }),
});
```

### `precheckFailSchema`

This Zod schema defines the structure of a failed `precheck` result. What's included in the returned object is up to you, but ideally it includes details about why the `precheck` failed.

The following schema returns additional information to the Vincent Tool executor that would help them understand why the tool execution would fail. In this case, the `reason` string allows the `precheck` function to return a specific error message stating something like `"Insufficient token balance"` or `"Insufficient native token balance"`, along with current and required amounts for debugging:

> > **Note** If any unhandled error occurs during execution of your tool's `precheck` function, the Vincent Tool & Policy SDK will automatically return a `fail` result with the error message.

Reviewer: Consider adding
```suggestion
> **Note:** If any unhandled error occurs during execution of your tool's `precheck` function, the Vincent Tool & Policy SDK will automatically return a `fail` result with the error message.
```

  }),
});
```

## Execute Function

The `execute` function is the main logic of your Vincent Tool, executed within the Lit Action environment when the Vincent Tool executor wants to perform the actual tool operation on behalf of the Vincent App User.

Unlike the `precheck` function which only validates feasibility, the `execute` function performs the actual work your tool is designed to do. Additionally, because the `execute` function is executed in the Lit Action environment, it has access to the full Lit Action capabilities, including the ability to sign transactions and data using the Vincent App User's Agent Wallet (for more information on what's available to you within the Lit Action environment see the Lit Protocol [Lit Action](https://developer.litprotocol.com/sdk/serverless-signing/overview) docs).

> **Note** This [Lit Action doc page](https://developer.litprotocol.com/sdk/serverless-signing/combining-signatures) covers how to sign data with a PKP using the Ethers.js library within a Lit Action. Ethers.js is injected by Lit into the Lit Action runtime, so you don't need to import it to use it within your tool's `execute` function.

> Before executing your tool's `execute` function, the Vincent Tool & Policy SDK will execute the `evaluate` functions of the Vincent Policies enabled by the Vincent App User for your tool for the specific Vincent App the tool is being executed for. If all Vincent Policies return `allow` results, your tool's `execute` function will be executed.

Reviewer: Consider reducing this to
```suggestion
Before executing your tool's `execute` function, the Vincent Tool & Policy SDK will execute the `evaluate` functions of the Vincent Policies. If all Vincent Policies return `allow` results, your tool's `execute` function will be executed.
```

Author: Changed to:

Before executing your tool's `execute` function, the Vincent Tool & Policy SDK will execute the `evaluate` functions of any Vincent Policies registered by the Vincent User. If all policies return `allow` results, your tool's `execute` function will be executed.

The following schema returns to the Vincent Tool executor the transaction hash of the executed transaction, and the hash for the transaction sent during the execution of the Vincent spending limit policy's `commit` function to update the amount of tokens spent (this is covered further in the [Executing Vincent Policy Commit Functions](#executing-vincent-policy-commit-functions) section):

```typescript
import { createVincentTool } from '@lit-protocol/vincent-tool-sdk';
import { z } from 'zod';

const vincentTool = createVincentTool({
  // ... other policy definitions

  executeSuccessSchema: z.object({
    transferTransactionHash: z.string(),
    spendTransactionHash: z.string().optional(),
  }),
});
```

### `executeFailSchema`

This Zod schema defines the structure of a failed `execute` result. What's included in the returned object is up to you, but ideally it includes details about why the `execute` function failed.

The following schema returns error information to the Vincent Tool executor, including an error message, error code, and revert reason for failed transactions to assist with debugging:

> > **Note** If any unhandled error occurs during execution of your tool's `execute` function, the Vincent Tool & Policy SDK will automatically return a `fail` result with the error message.

Reviewer: Consider adding
```suggestion
> **Note:** If any unhandled error occurs during execution of your tool's `execute` function, the Vincent Tool & Policy SDK will automatically return a `fail` result with the error message.
```

> **WARNING** Your tool's `execute` function should always call the `commit` functions for any of your tool's supported Vincent Policies that have a `commit` function defined. Failing to do so could cause the Vincent Policies to operate incorrectly, failing to deny tool executions that exceed the Vincent App User's defined limits.

Vincent Policy commit functions are **optional** - not all policies will have them. They're typically used by policies that need to track cumulative data like spending amounts, execution counts, or other stateful information that depends on successful tool execution.

After all the Vincent Policies that have been registered to be used with your tool for the specific Vincent App the tool is being executed for have been evaluated, the `policiesContext` property from the `toolContext` object will be updated to contain the policy evaluation results.

The `policiesContext` object contains a property called `allowedPolicies` that is an object where the keys are the package names of the evaluated Vincent policies, and the values are objects containing the `evalAllowResult` of the policy, and the policy's `commit` function if one exists for the policy:

> > **Note** The following interface isn't the actual interface used by the Vincent Tool & Policy SDK, it's just a simplified example of what the `policiesContext` object looks like for reference.

Reviewer: Consider adding
```suggestion
> **Note:** The following interface isn't the actual interface used by the Vincent Tool & Policy SDK, it's just a simplified example of what the `policiesContext` object looks like for reference.
```
